File: api-app/src/main/java/de/muspellheim/activitysampling/api/domain/activities/LogActivityCommand.java
// Copyright (c) 2025 Falko Schumann. All rights reserved. MIT license.

package de.muspellheim.activitysampling.api.domain.activities;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import java.time.Duration;
import java.time.Instant;
import lombok.Builder;
import lombok.With;

@Builder
@With
@JsonInclude(Include.NON_NULL)
public record LogActivityCommand(
    @NotNull Instant timestamp,
    @NotNull Duration duration,
    @NotBlank String client,
    @NotBlank String project,
    @NotBlank String task,
    String notes) {

  public static LogActivityCommand createTestInstance() {
    return LogActivityCommand.builder()
        .timestamp(Instant.parse("2024-12-18T08:30:00Z"))
        .duration(Duration.ofMinutes(30))
        .client("Test client")
        .project("Test project")
        .task("Test task")
        .build();
  }
}

File: src/shared/domain/logged-activity/log_activity.command.ts
// Copyright (c) 2026 Falko Schumann. All rights reserved. MIT license.

export class LogActivityCommand {
  static create({
    timestamp,
    duration,
    client,
    project,
    task,
    notes,
    category,
  }: {
    timestamp: Temporal.Instant | string;
    duration: Temporal.DurationLike | string;
    client: string;
    project: string;
    task: string;
    notes?: string;
    category?: string;
  }) {
    return new LogActivityCommand(
      timestamp,
      duration,
      client,
      project,
      task,
      notes,
      category,
    );
  }

  static createTestInstance({
    timestamp = "2025-08-14T11:00:00Z",
    duration = "PT30M",
    client = "Test client",
    project = "Test project",
    task = "Test task",
    notes,
    category,
  }: {
    timestamp?: Temporal.Instant | string;
    duration?: Temporal.DurationLike | string;
    client?: string;
    project?: string;
    task?: string;
    notes?: string;
    category?: string;
  } = {}) {
    return LogActivityCommand.create({
      timestamp,
      duration,
      client,
      project,
      task,
      notes,
      category,
    });
  }

  readonly type = "log-activity";
  readonly data;

  private constructor(
    timestamp: Temporal.Instant | string,
    duration: Temporal.DurationLike | string,
    client: string,
    project: string,
    task: string,
    notes?: string,
    category?: string,
  ) {
    this.data = {
      timestamp: Temporal.Instant.from(timestamp),
      duration: Temporal.Duration.from(duration),
      client: client,
      project: project,
      task: task,
      notes: notes,
      category: category,
    };
  }
}
